### ISSUE TITLE
about the prediction

### ISSUE THREAD

**Issue by xinyooo** [jaywonchung/BERT4Rec-VAE-Pytorch#6]
Hi, I used your bert4rec model to predict a sequence, but I don't know how to let the tensor readable. Can you give me some tips? thx!
Best regards

        # load one seq
        seq = dataset['test'][i]
        mask_token = len(dataset['smap']) + 1
        seq = seq + [mask_token]
        seq = seq[-args.bert_max_len:]
        padding_len = args.bert_max_len - len(seq)
        seq = [0] * padding_len + seq
        seq_ = torch.LongTensor([seq]).to(device)
        # recommend 50 item
        rec_list = []
        for _ in range(50):
            print(seq)
            #input 1...T time seq to predict
            logit = model(seq_)[0]
            #get the last tensor which is prediction? (T+1) time
            rec = torch.argmax(logit[-1], 0).it em()
            rec_list.append(datamap[rec])
            seq.insert(-1, rec)
            seq.pop(0)
            seq_ = torch.LongTensor([seq]).to(device)
            break

**Comment by xinyooo** [comment]
input seq is:
`[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 12978, 1060, 4591, 7851, 8655, 6747, 480, 6960, 17536]`
and the output seq's tensor is:
`tensor([[-6.0399, -6.0244,  1.3795,  ..., -3.5666, -6.0616, -6.0487],
        [-6.5003, -6.4871, -0.0961,  ..., -0.2114, -6.5191, -6.5089],
        [-6.1263, -6.1145,  0.9033,  ..., -0.9187, -6.1415, -6.1320],
        ...,
        [-6.1947, -6.1808, -3.1550,  ..., -1.7596, -6.2164, -6.2032],
        [-6.7190, -6.7046, -0.5234,  ...,  0.4454, -6.7387, -6.7288],
        [-6.4853, -6.4697, -0.1267,  ..., -0.2318, -6.5050, -6.4940]],
       device='cuda:0', grad_fn=<SelectBackward>)`

**Comment by jaywonchung** [comment]
Hi xinyooo,

I can't really thoroughly understand your prediction code, but it seems that you're making predictions in an auto-regressive manner. However, BERT4Rec, as is BERT, is not an auto-regressive model. Rather, it receives a list of items and directly outputs the prediction probabilities for all items. You can see this in the [`models/bert.py`](https://github.com/jaywonchung/BERT4Rec-VAE-Pytorch/blob/f66f2534ebfd937778c7174b5f9f216efdebe5de/models/bert.py#L7) class definition, where the `forward` method returns the output of a linear layer whose output dimension is `args.num_items + 1`.

Hope this helped.

Regards,
Jaewon

**Comment by xinyooo** [comment]
set max_len = 50
num_item = 80257

1. I give a list of item, padding it to length 50, list the seq and convert to LongTensor, the input seq shape is [1, 50]

17536 is mask
`
tensor([[    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0, 12978,  1060,  4591,  7851,  8655,  6747,   480,  6960, 17536]],
       device='cuda:0')
`
2. input to model, the model output the tensor of shape is  [1, 50, 80258]
`
tensor([[-6.0399, -6.0244,  1.3795,  ..., -3.5666, -6.0616, -6.0487],
        [-6.5003, -6.4871, -0.0961,  ..., -0.2114, -6.5191, -6.5089],
        [-6.1263, -6.1145,  0.9033,  ..., -0.9187, -6.1415, -6.1320],
        ...,
        [-6.1947, -6.1808, -3.1550,  ..., -1.7596, -6.2164, -6.2032],
        [-6.7190, -6.7046, -0.5234,  ...,  0.4454, -6.7387, -6.7288],
        [-6.4853, -6.4697, -0.1267,  ..., -0.2318, -6.5050, -6.4940]],
       device='cuda:0', grad_fn=<SelectBackward>)
`
3. get the last index of output tensor which is mask output `[-6.4853, -6.4697, -0.1267,  ..., -0.2318, -6.5050, -6.4940]]` the dims is 80258
4. then get the index of max value of probability in 80258 dims which is the recommend item
5. input the rec. item to seq then pop first element which is T = 1
6. now the seq is 
`
tensor([[         0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0, 12978,  1060,  4591,  7851,  8655,  6747,   480,  6960,  mapped(new REC.), 17536,]],
       device='cuda:0')
`
6. repeat for 50 times

It is work? because my prediction in this way always output same thing
